File: seminar-07/heap_sort/cpp/heap_sort_inplace.cpp
#include <vector>
#include <initializer_list>
#include <optional>
#include <utility>

template <class T>
class MaxHeap {
public:
    std::vector<T> a;

    MaxHeap() = default;

    MaxHeap(const std::vector<T>& data) {
        a = data;
        heapify();
    }

    MaxHeap(std::initializer_list<T> init) {
        a = init;
        heapify();
    }

    size_t size() const {
        return a.size();
    }

    bool is_empty() const {
        return a.empty();
    }

    std::optional<T> peek() const {
        if (a.empty()) return std::nullopt;
        return a[0];
    }

    void push(const T& x) {
        a.push_back(x);
        shift_up((int)a.size() - 1);
    }

    std::optional<T> pop() {
        if (a.empty()) return std::nullopt;
        T top = a[0];
        T last = a.back();
        a.pop_back();
        if (!a.empty()) {
            a[0] = std::move(last);
            shift_down(0);
        }
        return top;
    }

    void shift_up(int i) {
        while (i > 0) {
            int p = (i - 1) / 2;
            if (a[p] >= a[i]) break;
            std::swap(a[p], a[i]);
            i = p;
        }
    }

    void shift_down(int i, int n = -1) {
        if (n == -1) n = (int)a.size();
        while (true) {
            int l = 2 * i + 1;
            if (l >= n) break;
            int r = l + 1;
            int j = l;
            if (r < n && a[r] > a[l]) j = r;
            if (a[i] >= a[j]) break;
            std::swap(a[i], a[j]);
            i = j;
        }
    }

    void heapify() {
        int n = (int)a.size();
        for (int i = n / 2 - 1; i >= 0; --i) {
            shift_down(i, n);
        }
    }
};

template <class T>
void heap_sort_inplace(std::vector<T>& arr) {
    MaxHeap<T> heap(arr);
    arr = heap.a;

    int n = (int)arr.size();
    for (int end = n - 1; end > 0; --end) {
        std::swap(arr[0], arr[end]);
        heap.shift_down(0, end);
    }
}

File: seminar-07/heap_sort/python/heap_sort_inplace.py
class MaxHeap:
    def __init__(self, data=None):
        self.a = []
        if data is not None:
            self.a = list(data)
            self.heapify()

    def __len__(self):
        return len(self.a)

    def is_empty(self):
        return len(self.a) == 0

    def peek(self):
        if not self.a:
            return None
        return self.a[0]

    def push(self, x):
        self.a.append(x)
        self.shift_up(len(self.a) - 1)

    def pop(self):
        if not self.a:
            return None
        top = self.a[0]
        last = self.a.pop()
        if self.a:
            self.a[0] = last
            self.shift_down(0)
        return top

    def shift_up(self, i):
        a = self.a
        while i > 0:
            p = (i - 1) // 2
            if a[p] >= a[i]:
                break
            a[p], a[i] = a[i], a[p]
            i = p

    def shift_down(self, i, n=None):
        a = self.a
        if n is None:
            n = len(a)
        while True:
            l = 2 * i + 1
            if l >= n:
                break
            r = l + 1
            j = l
            if r < n and a[r] > a[l]:
                j = r
            if a[i] >= a[j]:
                break
            a[i], a[j] = a[j], a[i]
            i = j

    def heapify(self):
        n = len(self.a)
        for i in range(n // 2 - 1, -1, -1):
            self.shift_down(i, n)
            
def heap_sort_inplace(arr):
    heap = MaxHeap(arr)   # строим max-heap
    arr[:] = heap.a       # берем внутренний массив кучи

    n = len(arr)

    # извлекаем максимум в конец массива
    for end in range(n - 1, 0, -1):
        arr[0], arr[end] = arr[end], arr[0]   # максимум на свое место
        heap.shift_down(0, end)               # восстановить кучу на префиксе